src/multitaper.jl

```diff
@@ -0,0 +1,604 @@
+"""
+function mt_cross_spectral(signal::AbstractMatrix{T}; fs=1, kwargs...) where {T}
+    n_channels, n_samples = size(signal)
+    config = MTCrossSpectraConfig{T}(n_channels, n_samples; fs=fs, fft_flags=FFTW.ESTIMATE,
+                                     kwargs...)
+    output = allocate_output(config)
+    return mt_cross_spectral!(output, signal, config)
+end
+
+#####
+##### Multitapered coherences
+#####
+
+struct MTCoherenceConfig{T,T1,C<:MTCrossSpectraConfig{T}}
+    cs_matrix::T1
+    cs_config::C
+end
+
+"""
+    MTCoherenceConfig(cs_config::MTCrossSpectraConfig{T}) where {T}
+    MTCoherenceConfig{T}(n_channels, n_samples; fs=1, demean=false, low_bias=true, freq_range=nothing, kwargs...) where T
+
+Creates a configuration object for coherences from a [`MTCrossSpectraConfig`](@ref). Provides a helper method
+with the same arugments as `MTCrossSpectraConfig` to construct the `MTCrossSpectraConfig` object.
+"""
+function MTCoherenceConfig(cs_config::MTCrossSpectraConfig{T}) where {T}
+    cs_matrix = allocate_output(cs_config)
+    return MTCoherenceConfig{T,typeof(cs_matrix),typeof(cs_config)}(cs_matrix, cs_config)
+end
+
+# add a method to cover the case in which the user specifies the `{T}` here
+MTCoherenceConfig{T}(cs_config::MTCrossSpectraConfig{T}) where {T} = MTCoherenceConfig(cs_config)
+
+function MTCoherenceConfig{T}(n_channels, n_samples; fs=1, demean=false, low_bias=true,
+                              freq_range=nothing, kwargs...) where {T}
+    cs_config = MTCrossSpectraConfig{T}(n_channels, n_samples; fs=fs, demean=demean,
+                                        low_bias=low_bias, freq_range=freq_range, kwargs...)
+    cs_matrix = allocate_output(cs_config)
+    return MTCoherenceConfig{T,typeof(cs_matrix),typeof(cs_config)}(cs_matrix, cs_config)
+end
+
+function allocate_output(config::MTCoherenceConfig{T}) where {T}
+    return Array{real(T)}(undef, config.cs_config.n_channels, config.cs_config.n_channels, length(config.cs_config.freq))
+end
+
+"""
+    coherence_from_cs!(output::AbstractArray, cs_matrix)
+
+Compute the pairwise channel coherences from a cross spectral matrix, storing the result in `output`.
+"""
+function coherence_from_cs!(output::AbstractArray{T}, cs_matrix) where T
+    n_channels = size(cs_matrix, 2)
+    n_freqs = size(cs_matrix, 3)
+    @boundscheck checkbounds(output, 1:n_channels, 1:n_channels, 1:n_freqs)
+    @boundscheck checkbounds(cs_matrix, 1:n_channels, 1:n_channels, 1:n_freqs)
+    output .= zero(T)
+    @inbounds for ch2 in 1:n_channels
+        for ch1 in (ch2 + 1):n_channels # lower triangular matrix
+            for f in 1:n_freqs # average over frequency
+                output[ch1, ch2, f] += real(abs(cs_matrix[ch1, ch2, f]) /
+                                         (sqrt(cs_matrix[ch1, ch1, f] *
+                                               cs_matrix[ch2, ch2, f])))
+            end
+        end
+    end
+    output .+= PermutedDimsArray(output, (2, 1, 3)) # symmetrize
```

Comment:
we don't have a `Symmetric` type like in the matrix case, so I just symmetrize it by hand. If I haven't messed things up this should at least be non-allocating, and probably the expensive bit is all the FFTs anyway, so I think it's probably not that bad to copy in the right values here.

Reply:
Typically wouldn't you divide by 2 when symmetrizing this way?

Reply:
Not here because we only fill in the lower half (and not the diagonal), and leave the upper half zero. So then when we add the “transpose” here it just copies the lower half up to the upper half. (Then we set the diagonal to 1).